104 avatar view

GeorgeLeithead opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->

 ### Description of Change ###

**AvatarView** is a control for displaying a user's avatar image or their initials. The **AvatarView** supports text, border, and background colors as well as shaped avatars and supports shadow and gestures.

Inheriting from the **Border** control, with added properties for _BorderColor_, _BorderWidth_, .NET MAUI _CornerRadius_, _Font_, _ImageSource_, _Text_, _TextColor_ and _TextTransformation_ properties.  Further inheriting from IFontElement, ITextElement, IImageElement, ITextAlignmentElement, ILineHeightElement and ICornerElement.   

 ### Linked Issues ###
 #104 [Proposal] AvatarView
 - Fixes #

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear, please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/117 <!-- Replace XYZ with your docs PR #. The checkbox will be automatically checked once the docs PR has been merged -->


### Additional information ###
 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->
#### Platforms Tested ####
 - [x] Android
 - [x] Windows
 - [x] iOS
 - [x] MACCatalyst

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+public class AvatarModel
+{
+	public Color? AvatarBackgroundColor { get; set; } = Colors.Black;
```
Please make this Property read-only and remove nullability

```suggestion
	public Color AvatarBackgroundColor { get; init; } = Colors.Black;
```

GeorgeLeithead replied:
Fully understand and will do so.

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+public class AvatarModel
+{
+	public Color? AvatarBackgroundColor { get; set; } = Colors.Black;
+	public FontAttributes? FontAttributes { get; set; }
```
Please make this Property read-only
```suggestion
	public FontAttributes? FontAttributes { get; init; }
```

GeorgeLeithead replied:
Fully understand and will do so.

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+	public Color? AvatarBackgroundColor { get; set; } = Colors.Black;
+	public FontAttributes? FontAttributes { get; set; }
+
+	public double? FontSize { get; set; } = Device.GetNamedSize(NamedSize.Large, typeof(Label));
```
Please make this Property read-only and remove nullability
```suggestion
	public double FontSize { get; init; } = Device.GetNamedSize(NamedSize.Large, typeof(Label));
```

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public double? FontSize { get; set; } = Device.GetNamedSize(NamedSize.Large, typeof(Label));
+
+	public ImageSource? ImageSource { get; set; }
```
Please make this Property read-only and remove nullability 

```suggestion
	public ImageSource ImageSource { get; init; }
```

GeorgeLeithead replied:
Fully understand and will do so.

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public ImageSource? ImageSource { get; set; }
+
+	public string? Text { get; set; } = "?";
```
Please make this Property read-only and remove nullability 

```suggestion
	public string Text { get; init; } = "?";
```

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public string? Text { get; set; } = "?";
+
+	public Color? TextColor { get; set; } = Colors.White;
```
Please make this Property read-only and remove nullability 
```suggestion
	public Color TextColor { get; init; } = Colors.White;
```

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public Color? TextColor { get; set; } = Colors.White;
+
+	public Thickness? BorderWidth { get; set; } = 0;
```
Please make this Property read-only and remove nullability 
```suggestion
	public Thickness BorderWidth { get; init; } = 0;
```

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public Thickness? BorderWidth { get; set; } = 0;
+
+	public Thickness? AvatarPadding { get; set; } = 5;
```
Please make this Property read-only and remove nullability 
```suggestion
	public Thickness AvatarPadding { get; init; } = 5;
```

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public Thickness? AvatarPadding { get; set; } = 5;
+
+	public double? AvatarWidthRequest { get; set; } = 64;
```
Please make this Property read-only and remove nullability 
```suggestion
	public double AvatarWidthRequest { get; init; } = 64;
```

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public double? AvatarWidthRequest { get; set; } = 64;
+
+	public double? AvatarHeightRequest { get; set; } = 64;
```
Please make this Property read-only and remove nullability 
```suggestion
	public double AvatarHeightRequest { get; init; } = 64;
```

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public double? AvatarHeightRequest { get; set; } = 64;
+
+	public CornerRadius? CornerRadius { get; set; } = new(50, 50, 50, 50);
```
Please make this Property read-only and remove nullability 
```suggestion
	public CornerRadius CornerRadius { get; init; } = new(50, 50, 50, 50);
```

brminnick commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+
+	public CornerRadius? CornerRadius { get; set; } = new(50, 50, 50, 50);
+
+	public Color? BorderColor { get; set; }
```
Please make this Property read-only 
```suggestion
	public Color? BorderColor { get; init; }
```

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+using Microsoft.Maui.Controls.Internals;
+using Microsoft.Maui.Controls.Shapes;
+
+namespace CommunityToolkit.Maui.Controls;
```
Please move this file to `/Views/` folder

GeorgeLeithead replied:
I have moded this file to the /Views/ folder and renamed to AvatarView.cs. I believe I have resolved this issue.

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+namespace CommunityToolkit.Maui.Controls;
+
+/// <summary>Avatar content view.</summary>
+public class Avatar : ContentView, IAvatarElement, IFontElement, ITextElement, IImageElement
```
Please name this `AvatarView`

```suggestion
public class AvatarView : ContentView, IAvatarElement, IFontElement, ITextElement, IImageElement
```

GeorgeLeithead replied:
I'm not personally 100% that it is necessary or right to add "view" to the name of the control.  This is more like the "Button" or "Label" control but understand that it is also a member of the "Present data" subclass.  I will make the change if it is agreed to call this "AvatarView".

GeorgeLeithead replied:
I have renamed to AvatarView. I believe I have resolved this issue.

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+using System.ComponentModel;
```
Please organize elements in this file according to the following StyleCop rules:

```
According to the [StyleCop  Rules Documentation][1] the ordering is as follows.

Within a class, struct or interface: (SA1201 and SA1203)

 - Constant Fields
 - Fields
 - Constructors
 - Finalizers (Destructors)
 - Delegates
 - Events
 - Enums
 - Interfaces (*interface implementations*)
 - Properties
 - Indexers
 - Methods
 - Structs
 - Classes

Within each of these groups order by access: (SA1202)

 - public
 - internal
 - protected internal
 - protected
 - private

Within each of the access groups, order by static, then non-static: (SA1204)

 - static
 - non-static

Within each of the static/non-static groups of fields,  order by readonly, then non-readonly : (SA1214 and SA1215)

 - readonly
 - non-readonly

An unrolled list is 130 lines long, so I won't unroll it here. The methods part unrolled is:

 - public static methods
 - public methods
 - internal static methods
 - internal methods
 - protected internal static methods
 - protected internal methods
 - protected static methods
 - protected methods
 - private static methods
 - private methods
```

https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1201.md

GeorgeLeithead replied:
I don't currently use StyleCop, but will install the extension and endeavour to meet the standards.

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+public class Avatar : ContentView, IAvatarElement, IFontElement, ITextElement, IImageElement
+{
+	/// <summary>The backing store for the <see cref="IAvatarElement.AvatarBackgroundColor" /> bindable property.</summary>
+	public static readonly BindableProperty AvatarBackgroundColorProperty = BindableProperty.Create(nameof(IAvatarElement.AvatarBackgroundColor), typeof(Color), typeof(IAvatarElement), null, propertyChanged: OnBackgroundColorChanged);
```
Let's use `Colors.Transparent` instead of `null` for the default value because the `AvatarBackgroundColor` property is not nullable.

Let's also use `nameof(AvatarBackgroundColor)` instead of `nameof(IAvatarElement.AvatarBackgroundColor)` to reference the property in this class.

```suggestion
	public static readonly BindableProperty AvatarBackgroundColorProperty = BindableProperty.Create(nameof(AvatarBackgroundColor), typeof(Color), typeof(IAvatarElement), Colors.Transparant, propertyChanged: OnBackgroundColorChanged);
```

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+	public static readonly BindableProperty AvatarHeightRequestProperty = BindableProperty.Create(nameof(IAvatarElement.AvatarHeightRequest), typeof(double), typeof(VisualElement), defaultValue: AvatarElement.DefaultHeightRequest, propertyChanged: OnHeightRequestPropertyChanged);
+
+	/// <inheritdoc/>
+	public static readonly BindableProperty AvatarPaddingBottomProperty = BindableProperty.Create("AvatarPaddingBottom", typeof(double), typeof(IPaddingElement), default(double), propertyChanged: OnPaddingBottomChanged);
+
+	/// <inheritdoc/>
+	public static readonly BindableProperty AvatarPaddingLeftProperty = BindableProperty.Create("AvatarPaddingLeft", typeof(double), typeof(IPaddingElement), default(double), propertyChanged: OnPaddingLeftChanged);
+
+	/// <summary>The backing store for the <see cref="AvatarPadding" /> bindable property.</summary>
+	public static readonly BindableProperty AvatarPaddingProperty = BindableProperty.Create(nameof(IPaddingElement.Padding), typeof(Thickness), typeof(IPaddingElement), default(Thickness), propertyChanged: OnPaddingPropertyChanged, defaultValueCreator: PaddingDefaultValueCreator);
+
+	/// <inheritdoc/>
+	public static readonly BindableProperty AvatarPaddingRightProperty = BindableProperty.Create("AvatarPaddingRight", typeof(double), typeof(IPaddingElement), default(double), propertyChanged: OnPaddingRightChanged);
+
+	/// <inheritdoc/>
+	public static readonly BindableProperty AvatarPaddingTopProperty = BindableProperty.Create("AvatarPaddingTop", typeof(double), typeof(IPaddingElement), default(double), propertyChanged: OnPaddingTopChanged);
```
Do we actually need 5 separate properties for the Padding? Can we just use one `public Thickness AvatarPadding` property?

GeorgeLeithead replied:
I took this from Microsoft.Maui.Controls.PaddingElement.cs which implements all 5 separate properties for padding.  I have no problem in implementing one public Thickness AvatarPadding property, if you think it's the right way to go.

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+	public static readonly BindableProperty AvatarPaddingTopProperty = BindableProperty.Create("AvatarPaddingTop", typeof(double), typeof(IPaddingElement), default(double), propertyChanged: OnPaddingTopChanged);
+
+	/// <summary>The backing store for the <see cref="AvatarShadow" /> bindable property.</summary>
+	public static readonly BindableProperty AvatarShadowProperty = BindableProperty.Create(nameof(IAvatarElement.AvatarShadow), typeof(Shadow), typeof(VisualElement), defaultValue: null, propertyChanged: OnShadowPropertyChanged);
```
You specify the default value of `AvatarShadow` as `null` here, but `public Shadow AvatarShadow` is not nullable.

Should the property be nullable, or should its default value be something other than `null`?

GeorgeLeithead replied:
I took this from Microsoft.Maui.Controls.VisualElement.impl.cs.  I know that the M$FT code is subject to change, but I thought it best to use exactly as implemented by this repository.  Happy to go in whatever direction you feel is appropriate.

GeorgeLeithead replied:
AvatarShadow property has been removed.

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+	/// <inheritdoc/>
+	public static readonly BindableProperty AvatarPaddingRightProperty = BindableProperty.Create("AvatarPaddingRight", typeof(double), typeof(IPaddingElement), default(double), propertyChanged: OnPaddingRightChanged);
+
+	/// <inheritdoc/>
```
What is this XML inheriting from?

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+	/// <summary>The backing store for the <see cref="AvatarPadding" /> bindable property.</summary>
+	public static readonly BindableProperty AvatarPaddingProperty = BindableProperty.Create(nameof(IPaddingElement.Padding), typeof(Thickness), typeof(IPaddingElement), default(Thickness), propertyChanged: OnPaddingPropertyChanged, defaultValueCreator: PaddingDefaultValueCreator);
+
+	/// <inheritdoc/>
```
What is this XML inheriting from?

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+	/// <summary>The backing store for the <see cref="IAvatarElement.AvatarHeightRequest" /> bindable property.</summary>
+	public static readonly BindableProperty AvatarHeightRequestProperty = BindableProperty.Create(nameof(IAvatarElement.AvatarHeightRequest), typeof(double), typeof(VisualElement), defaultValue: AvatarElement.DefaultHeightRequest, propertyChanged: OnHeightRequestPropertyChanged);
+
+	/// <inheritdoc/>
```
What is this XML inheriting from?

GeorgeLeithead replied:
My apologies.  This was developer laziness and I failed to go back and implement correctly.  Will implement correctly.

GeorgeLeithead replied:
I believe I have resolved this issue.

brminnick commented on src/CommunityToolkit.Maui/Controls/Avatar.cs:
```diff
@@ -0,0 +1,557 @@
+	/// <inheritdoc/>
+	public static readonly BindableProperty AvatarPaddingBottomProperty = BindableProperty.Create("AvatarPaddingBottom", typeof(double), typeof(IPaddingElement), default(double), propertyChanged: OnPaddingBottomChanged);
+
+	/// <inheritdoc/>
```
What is this XML inheriting from?

GeorgeLeithead replied:
My apologies. This was developer laziness and I failed to go back and implement correctly. Will implement correctly.

brminnick commented on src/CommunityToolkit.Maui/Controls/IAvatarElement.cs:
```diff
@@ -0,0 +1,111 @@
+
+/// <summary>Avatar content view element interface.</summary>
+[EditorBrowsable(EditorBrowsableState.Never)]
+public interface IAvatarElement
```
- Please move this file to the `CommunityToolkit.Maui.Core` project into its `Interfaces` folder
- Please change the namespace to `CommunityToolkit.Maui.Core`
- Please rename this to  `IAvatarView`
- Please have `IAvatarView` inherit from `IContentView, IBorder, IFont, IText, Microsoft.Maui.IImage`
```suggestion
public interface IAvatarView : IContentView, IBorder, IFont, IText, IImageSource, Microsoft.Maui.IImageSourcePart
```
  - Please then remove duplicate properties like `string Text { get; }` which is inherited from `IText`

GeorgeLeithead replied:
Fully understand and will move as indicated, change the namespace, rename the interface and have it inherit as appropriate.

brminnick commented on src/CommunityToolkit.Maui/Controls/AvatarElement.cs:
```diff
@@ -0,0 +1,23 @@
+﻿namespace CommunityToolkit.Maui.Controls;
```
- Please rename this file to `AvatarViewDefaults`
- Please move this file to the `CommunityToolkit.Maui.Core` project under the `/Primatives/Defaults/` folder

GeorgeLeithead replied:
Fully understand.  I will move and rename as indicated.

GeorgeLeithead replied:
I have renamed this file to AvatarViewDefault.shared.cs
I have moved this file to the CommunityToolKit.Maui.Core project under the /Primatives/Defaults/ folder.

brminnick reviewed (changes requested):
Thanks @GeorgeLeithead!

Please make minimal changes to existing files, and avoid formatting changes that don't improve functionality; `MauiProgram.cs` for example is flooded with unnecessary formatting changes making our job as reviewers much more difficult.

There are some larger changes need to be made at the architectural level:
- Rename `AvatarElement` -> `AvatarViewDefaults` + move it to `CommunityToolkit.Maui.Core` in the `/Primatives/Defaults` folder
- Rename `IAvatarElement` -> `IAvatarView` + move it to `CommunityToolkit.Maui.Core` in the `/Interfaces/` folder
- Rename `Avatar` -> `AvatarView` + move it to `/Views/` folder

Then there are some smaller changes that need to be done at the file level:
**AvatarView**
- Some bindable properties have `null` default values, but their associated properties are not nullable
- `AvatarView` is implementing .NET MAUI Interfaces; these interfaces should be inherited by the `IAvatarView` interface
- Some XML Docs are missing (and instead are using `///<inheritdoc/>` despite there being nothing to inherit
- You have 5 `double` Properties for the Padding; let's reduce this to 1 `Thickness` property

**IAvatarView**
- Let's ensure `IAvatarView` should inherit from as many existing .NET MAUI interfaces as possible 

There's some more I'd like to suggest, but let's start with these to see how the `AvatarView` takes shape 👍

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/AppShell.xaml:
```diff
@@ -1,78 +1,104 @@
-       xmlns:pages="clr-namespace:CommunityToolkit.Maui.Sample.Pages"
-       xmlns:sys="clr-namespace:System;assembly=netstandard"
-       Padding="4,0,0,0">
+<Shell
```
Fully understand and will do so. Looks like some of my installed extensions made the changes automatically.  I thought I had caught them, but clearly not.

GeorgeLeithead replied:
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/AppShell.xaml:
```diff
@@ -1,78 +1,104 @@
-    <FlyoutItem Title="Converters"
-                Route="ConvertersGalleryPage"
-                Icon="{OnPlatform Default='dotnet_bot.png', MacCatalyst=''}">
+    <FlyoutItem
+        Title="Controls"
+        Icon="{OnPlatform Default='dotnet_bot.png',
+                          MacCatalyst=''}"
+        Route="ControlsGalleryPage">
+        <ShellContent ContentTemplate="{DataTemplate controls:ControlsGalleryPage}" />
```
Fully understand and will do so.

GeorgeLeithead replied:
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:
```diff
@@ -84,9 +86,15 @@ public partial class AppShell : Shell
 		CreateViewModelMapping<PopupPositionPage, PopupPositionViewModel, ViewsGalleryPage, ViewsGalleryViewModel>(),
+
+		// Add Controls View Models
+		CreateViewModelMapping<AvatarControlPage, AvatarControlViewModel, ControlsGalleryPage, ControlsGalleryViewModel>(),
```
Fully understand and will do so.

GeorgeLeithead replied:
I have moved the page to the ViewsGalleryPage. I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:
```diff
@@ -95,18 +103,21 @@ public static string GetPageRoute(Type viewModelType)
 		}
 
-		var uri = new UriBuilder("", GetPageRoute(mapping.GalleryPageType, mapping.ContentPageType));
+		UriBuilder uri = new("", GetPageRoute(mapping.GalleryPageType, mapping.ContentPageType));
```
Fully understand and will do so. Looks like some of my installed extensions made the changes automatically.  I thought I had caught them, but clearly not.

GeorgeLeithead replied:
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:
```diff
@@ -95,18 +103,21 @@ public static string GetPageRoute(Type viewModelType)
 	}
 
-	static string GetPageRoute(Type galleryPageType, Type contentPageType) => $"//{galleryPageType.Name}/{contentPageType.Name}";
+	static string GetPageRoute(Type galleryPageType, Type contentPageType)
+	{
+		return $"//{galleryPageType.Name}/{contentPageType.Name}";
+	}
```
Fully understand and will do so. Looks like some of my installed extensions made the changes automatically.  I thought I had caught them, but clearly not.

GeorgeLeithead replied:
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:
```diff
@@ -95,18 +103,21 @@ public static string GetPageRoute(Type viewModelType)
 																																							where TGalleryViewModel : BaseGalleryViewModel
 	{
-		var route = GetPageRoute(typeof(TGalleryPage), typeof(TPage));
+		string route = GetPageRoute(typeof(TGalleryPage), typeof(TPage));
```
Fully understand and will do so. Looks like some of my installed extensions made the changes automatically.  I thought I had caught them, but clearly not.

GeorgeLeithead replied:
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/MauiProgram.cs:
```diff
@@ -20,13 +22,13 @@ public static class MauiProgram
 	public static MauiApp CreateMauiApp()
 	{
-		var builder = MauiApp.CreateBuilder()
+		MauiAppBuilder builder = MauiApp.CreateBuilder()
```
Fully understand and will do so. Looks like some of my installed extensions made the changes automatically.  I thought I had caught them, but clearly not.

GeorgeLeithead replied:
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/MauiProgram.cs:
```diff
@@ -38,166 +40,174 @@ public static MauiApp CreateMauiApp()
+		_ = services.AddTransient<IsNullConverterPage>();
+		_ = services.AddTransient<IsStringNotNullOrEmptyConverterPage>();
+		_ = services.AddTransient<IsStringNotNullOrWhiteSpaceConverterPage>();
+		_ = services.AddTransient<IsStringNullOrEmptyConverterPage>();
+		_ = services.AddTransient<IsStringNullOrWhiteSpaceConverterPage>();
+		_ = services.AddTransient<ItemTappedEventArgsConverterPage>();
+		_ = services.AddTransient<ListToStringConverterPage>();
+		_ = services.AddTransient<MathExpressionConverterPage>();
+		_ = services.AddTransient<MultiConverterPage>();
+		_ = services.AddTransient<MultiMathExpressionConverterPage>();
+		_ = services.AddTransient<SelectedItemEventArgsConverterPage>();
+		_ = services.AddTransient<StateToBooleanConverterPage>();
+		_ = services.AddTransient<StringToListConverterPage>();
+		_ = services.AddTransient<TextCaseConverterPage>();
+		_ = services.AddTransient<VariableMultiValueConverterPage>();
 
 		// Add Extensions Pages
-		services.AddTransient<ColorAnimationExtensionsPage>();
+		_ = services.AddTransient<ColorAnimationExtensionsPage>();
 
 		// Add Layouts Pages
-		services.AddTransient<UniformItemsLayoutPage>();
+		_ = services.AddTransient<UniformItemsLayoutPage>();
 
 		// Add Views Pages
-		services.AddTransient<CsharpBindingPopup>();
-		services.AddTransient<DrawingViewPage>();
-		services.AddTransient<MultiplePopupPage>();
-		services.AddTransient<PopupAnchorPage>();
-		services.AddTransient<PopupPositionPage>();
-		services.AddTransient<XamlBindingPopup>();
+		_ = services.AddTransient<CsharpBindingPopup>();
+		_ = services.AddTransient<DrawingViewPage>();
+		_ = services.AddTransient<MultiplePopupPage>();
+		_ = services.AddTransient<PopupAnchorPage>();
+		_ = services.AddTransient<PopupPositionPage>();
+		_ = services.AddTransient<XamlBindingPopup>();
+
+		// Add Controls Pages
+		_ = services.AddTransient<AvatarControlPage>();
```
Fully understand and will do so. Looks like some of my installed extensions made the changes automatically.  I thought I had caught them, but clearly not.

GeorgeLeithead replied:
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/MauiProgram.cs:
```diff
@@ -38,166 +40,174 @@ public static MauiApp CreateMauiApp()
 	{
-		services.AddSingleton<IDeviceInfo>(DeviceInfo.Current);
-		services.AddSingleton<IDeviceDisplay>(DeviceDisplay.Current);
+		_ = services.AddSingleton<IDeviceInfo>(DeviceInfo.Current);
+		_ = services.AddSingleton<IDeviceDisplay>(DeviceDisplay.Current);
```
Fully understand and will do so. Looks like some of my installed extensions made the changes automatically.  I thought I had caught them, but clearly not.

GeorgeLeithead replied:
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/Pages/ViewControls/AvatarControlPage.xaml:
```diff
@@ -0,0 +1,285 @@
+    </pages:BasePage.Resources>
+
+    <ScrollView>
+        <VerticalStackLayout x:Name="ContainerControl" Spacing="12">
```
I was fighting with myself on just how to present lots of examples to the end user and went thru a lot of different options.  I'll change to a grid, and I intend to add lots of semantic properties to help further describe the samples (and show best accessibility practice).

GeorgeLeithead replied:
I have implemented as a Grid. I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/Pages/ViewControls/AvatarControlPage.xaml.cs:
```diff
@@ -0,0 +1,34 @@
+using System.Diagnostics;
```
Fully understand and move as indicated.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/Pages/ViewControls/AvatarControlPage.xaml.cs:
```diff
@@ -0,0 +1,34 @@
+		InitializeComponent();
+	}
+
+	public void OnAvatarClicked(object? sender, EventArgs e)
```
Fully understand and will change the access modifier.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/Pages/ViewControls/AvatarControlPage.xaml.cs:
```diff
@@ -0,0 +1,34 @@
+		Debug.WriteLine("Clicked");
+	}
+
+	public void TapGestureRecognizer_Tapped(object sender, EventArgs e)
```
Fully understand and will change the access modifier.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/Pages/ViewControls/ControlsGalleryPage.cs:
```diff
@@ -0,0 +1,11 @@
+﻿using CommunityToolkit.Maui.Sample.ViewModels.ViewControls;
```
Fully understand and will move/make the structure changes as indicated.

GeorgeLeithead replied:
I have removed the ViewControls folder and files. I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarControlViewModel.cs:
```diff
@@ -0,0 +1,14 @@
+/// <summary>Sample Avatar control view model.</summary>
+public sealed partial class AvatarControlViewModel : BaseViewModel
+{
+	public ObservableCollection<AvatarModel> AvatarList { get; }
+
+	public AvatarControlViewModel()
+	{
+		AvatarList = new ObservableCollection<AvatarModel>(AvatarData.Avatars);
+	}
```
Will make the appropriate change.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+{
+	public Color? AvatarBackgroundColor { get; set; } = Colors.Black;
+	public FontAttributes? FontAttributes { get; set; }
+
```
I understand this and appreciate it, however it is still used in Microsoft.MAUI and I believe it's an effective way to show the use of named font sizes.  Will make the change to a numeric font size if appropriate?

brminnick replied:
I'd prefer not to add code that leverages an `[Obesolete]` API because it'll cause a breaking change in our Toolkit in a future release of .NET MAUI.

I'll ping the team to see what the replacement for `Device.GetNamedSize` should be 👍

GeorgeLeithead replied:
I have removed these and replaced with numeric values.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarData.cs:
```diff
@@ -0,0 +1,89 @@
+				BorderWidth = 3,
+				AvatarPadding = 5,
+				AvatarWidthRequest = 48,
+				AvatarHeightRequest = 64,
+				FontSize = Device.GetNamedSize(NamedSize.Medium, typeof(Label)),
+				CornerRadius = new CornerRadius(16,16,0,0),
+				BorderColor = Colors.Pink,
+				FontAttributes = FontAttributes.Bold,
+			},
+
+			new AvatarModel
+			{
+				Text = "SL", // Stephen Lang
+				TextColor = Colors.White,
+				AvatarBackgroundColor = Colors.Black,
+				BorderWidth = Thickness.Zero,
+				AvatarPadding = Thickness.Zero,
+				AvatarWidthRequest = 64,
+				AvatarHeightRequest = 48,
+				FontSize = Device.GetNamedSize(NamedSize.Medium, typeof(Label)),
+				CornerRadius = new CornerRadius(0,0,16,16),
+				BorderColor = Colors.Red,
+				FontAttributes = FontAttributes.Italic,
+			},
+
+			new AvatarModel
+			{
+				Text = "Michelle Rodriguez", // Michelle Rodriguez
+				TextColor = Colors.Yellow,
+				AvatarBackgroundColor = Colors.Gray,
+				BorderWidth = 1,
+				AvatarPadding = 7,
+				FontSize = Device.GetNamedSize(NamedSize.Medium, typeof(Label)),
+				CornerRadius = new CornerRadius(0,0,16,16),
+				BorderColor = Colors.Orange,
+				FontAttributes = FontAttributes.Italic,
+				AvatarWidthRequest = 128,
+				AvatarHeightRequest = 72,
+			},
+		};
```
Fully understand and will do so.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarModel.cs:
```diff
@@ -0,0 +1,27 @@
+﻿namespace CommunityToolkit.Maui.Sample.ViewModels.ViewControls;
```
Fully understand and will do so. Initially I wanted to keep everything I was working on 'away' from everything else, to improve isolation whilst I worked on this and tried out different ideas.  As it's becoming more concrete, you are right, it is now better to follow approved standards.

GeorgeLeithead replied:
I have moved this file to the /Models/ folder. I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/ControlsGalleryViewModel.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using CommunityToolkit.Maui.Sample.Models;
```
Fully understand and will do so.

GeorgeLeithead replied:
I have deleted this file. I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:
```diff
@@ -84,9 +86,15 @@ public partial class AppShell : Shell
 	});
 
-	public AppShell() => InitializeComponent();
+	public AppShell()
+	{
+		InitializeComponent();
+	}
```
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/MauiProgram.cs:
```diff
@@ -38,166 +40,174 @@ public static MauiApp CreateMauiApp()
+		_ = services.AddTransient<XamlBindingPopupViewModel>();
+
+		// Add Controls View Models
+		_ = services.AddTransient<AvatarControlViewModel>();
```
I believe I have resolved this issue.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/ViewModels/ViewControls/AvatarControlViewModel.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections.ObjectModel;
```
I have moved this file to /ViewModels/Views/ folder and renamed to AvatarViewViewModel.cs

brminnick commented:
Hey George! 

Just FYI - I updated the implemented interfaces and removed some duplicate properties

```cs
public class AvatarView : Border, IAvatarView, IFontElement, ITextElement, IImageElement, ITextAlignmentElement, ILineHeightElement, ICornerElement
```
- Replaced `ContentView` -> `Border`
  - We were previously assigning `ContentView.Content = Border`
  - Replacing `ContentView` with `Border` removes this unnecessary layer and improves performance
- Add `IImageElement`, `ITextAlignmentElement`, `ILineHeightElement`, `ICornerElement`
- Removed `AvatarPadding*` properties
  - These aren't needed because `Border` already implements `public Thickness Padding`
- Removed `AvatarShadow` property
  - This isn't needed because `Border` already implements `public Shadow Shadow`


```cs
public interface IAvatarView : IBorderView, ILabel, Microsoft.Maui.IImage, IImageSource
```
- Replaced `IBorder` -> `IBorderView`
- Replaced `IText` -> `ILabel`
- Added `IImage`

GeorgeLeithead commented:
@bijington I'll add them as a checklist in a comment in the PR.  I think it's important to have these 'visible'.  For example, for the AvatarView, when using an ImageSource and a CornerRadius, the ImageSource isn't clipped to the CornerRadius.  If you change the CorenerRadius AFTER initial render (such as in debug/hot-reload), the imageSource is then correctly clipped on all corners APART from the bottom right!

GeorgeLeithead commented:
@brminnick @bijington
There are a few items I'd like to raise as discussion points, regarding AvatarView:

- Add new bindable property to provide 'Rotation' of text/image? This would allow for a 'pin' style AvatarView. See this sample, where the AvatarView has a Rotation property of 45:
![image](https://user-images.githubusercontent.com/43666574/177801420-ea556200-b36e-4d1d-8c3f-5b84c86cb315.png)
- Given the ScrollView has a bug on Android, should the sample page be changed to 'fit' a standard Android screen?
- Should we provide an enum (and bindable property) of standardised avatar sizes, which if set, overrides any WidthRequest & HeightRequest?
- Should we provide an enum (and bindable property) of standardised shapes? I'm not sure we should, as it is possible to create some geometric patterns by adjusting the corner radius!

GeorgeLeithead commented:
I'm also going to start working on the test cases, and then the documentation (not my strongest point).

bijington commented:
> I'm also going to start working on the test cases, and then the documentation (not my strongest point).

I'll be happy to assist with the docs writing if you need.

VladislavAntonyuk commented on samples/CommunityToolkit.Maui.Sample/ViewModels/Views/AvatarViewViewModel.cs:
```diff
@@ -0,0 +1,91 @@
+/// <summary>Sample view model for Avatar View page.</summary>
+public partial class AvatarViewViewModel : BaseViewModel
+{
+#pragma warning disable CS0612 // Type or member is obsolete, this is however still used in Microsoft.Maui
```
Let’s set the font size as a number, so we do not need to use obsolete methods

GeorgeLeithead replied:
@brminnick Is checking to see what to use, and I'll adjust accordingly  https://github.com/CommunityToolkit/Maui/pull/481#discussion_r915417622

VladislavAntonyuk commented on src/CommunityToolkit.Maui.Core/Interfaces/IAvatarView.shared.cs:
```diff
@@ -0,0 +1,67 @@
+	//// NOTE to implementer: Implement these methods explicitly
+
+	/// <summary>Gets a value indicating the corner radius default.</summary>
+	CornerRadius CornerRadiusDefaultValue { get; }
```
I think it should be in Public properties region. Do we need regions at all? As for me it’s a code smell

GeorgeLeithead replied:
@VladislavAntonyuk I copied the NOTES from Microsoft.Maui, and I added the regions to help differentiate.  I normally only ever use regions to segregate code when I'm working on it, but for production I tend to remove them.
I'll remove them.

VladislavAntonyuk commented on src/CommunityToolkit.Maui/Views/AvatarView.cs:
```diff
@@ -0,0 +1,333 @@
+using System.ComponentModel;
```
The file name should be AvatarView.shared.cs

GeorgeLeithead replied:
Renamed.

GeorgeLeithead commented on samples/CommunityToolkit.Maui.Sample/Models/AvatarModel.cs:
```diff
@@ -0,0 +1,30 @@
+	public FontAttributes? FontAttributes { get; init; }
+
+#pragma warning disable CS0612 // Type or member is obsolete, however is still used in Microsoft.Maui
+	public double FontSize { get; init; } = Device.GetNamedSize(NamedSize.Large, typeof(Label));
```
@brminnick Is checking to see what to use, and I'll adjust accordingly  https://github.com/CommunityToolkit/Maui/pull/481#discussion_r915417622

brminnick replied:
Hey George! Vlad is correct, we should assign a number for the FontSize and MAUI will automatically scale it:

https://github.com/dotnet/maui/issues/3938

GeorgeLeithead replied:
I'm changing these to numbers, whilst I work on implementing https://github.com/CommunityToolkit/Maui/pull/481#issuecomment-1177906589

GeorgeLeithead commented:
Still working on this push (https://github.com/CommunityToolkit/Maui/pull/481/commits/d75a3a2d0ed7298ba2958e6d7df851555fa484e6).  Still getting the styles sorted, but wanted to allow the community to see the changes.

GeorgeLeithead commented:
> > I'm also going to start working on the test cases, and then the documentation (not my strongest point).
> 
> I'll be happy to assist with the docs writing if you need.

@bijington I have written the documentation and submitted a **draft** pull request. [MicrosoftDocs/CommunityToolkit/pull/117](https://github.com/MicrosoftDocs/CommunityToolkit/pull/117)

brminnick reviewed (changes requested):
Thanks @GeorgeLeithead!! This is looking really good!!

Just FYI - I updated the Sample App, converting your popups into their own pages.

I found a couple things I'm hoping that you can look into for me:

### AvatarView Crashing on iOS, Android + MacCatalyst

When running on iOS, Android and MacCatalyst, any time I try to load an `AvatarView*Page` in the Sample App, the app crashes. Here's a gif demonstrating the behavior I'm seeing:
| iOS | Android |
| --   | -------- |
| ![ScreenFlow](https://user-images.githubusercontent.com/13558917/179336332-d8667f03-8ff7-46b3-a1cb-86032d6e493c.gif) | ![ScreenFlow2](https://user-images.githubusercontent.com/13558917/179336428-acf6b651-42e3-495a-9978-3fb0fc372672.gif) |

### AvatarImage Doesn't Clip to AvatarView Bounds
I noticed that the AvatarImage sticks out past the border of the AvatarView.

Let's make sure the image stays inside the bounds of the border.

<img width="1360" alt="Screen Shot 2022-07-15 at 7 14 36 PM" src="https://user-images.githubusercontent.com/13558917/179336447-c0ed0b80-0150-4df1-a474-cf349cd33e44.png">


### AvatarView Gestures Page

Let's make the gestures a bit more noticeable to the user. Instead of just outputting to the console, let's provide something visible to the user:
- **DragGestureRecognizer** - Allow the user to drag AvatarView across the screen
- **PinchGestureRecognizer** - Allow the user to resize the AvatarView height/width 
- **PanGestureRecognizer** - Allow the user to pan the AvatarView across the screen
- **SwipeGestureRecognizer** - Allow the user to swipe the AvatarView across the screen
- **TapGestureRecognizer** - Update the text in the AvatarLabel to the current tap count

GeorgeLeithead commented:
> ### AvatarImage Doesn't Clip to AvatarView Bounds
> I noticed that the AvatarImage sticks out past the border of the AvatarView.
> 
> Let's make sure the image stays inside the bounds of the border.
> 
> <img alt="Screen Shot 2022-07-15 at 7 14 36 PM" width="1360" src="https://user-images.githubusercontent.com/13558917/179336447-c0ed0b80-0150-4df1-a474-cf349cd33e44.png">

@brminnick The problem with the clip only happens on the Windows platform, and appears to be a bug in MAUI.  It is even more of an issue when you change the clip (in debug for example) after first render.  First render, no clip.  Change the top left, bottom left and top right and it'll clip these.  However, it will not clip bottom right!

GeorgeLeithead commented:
> Just FYI - I updated the Sample App, converting your popups into their own pages.
> 
> I found a couple things I'm hoping that you can look into for me:
> 
> ### AvatarView Crashing on iOS, Android + MacCatalyst
> When running on iOS, Android and MacCatalyst, any time I try to load an `AvatarView*Page` in the Sample App, the app crashes. 

@brminnick I have gone thru the new pages (thanks for this BTW) and cleaned them up, made them consistent, and verified that they do not crash on Windows or Android (I have found that I sometimes need to re-build when changing debug targets).  I can't verify on iOS or MacCatalyst as I don't have any such devices!

GeorgeLeithead commented:
> ### AvatarView Gestures Page
> Let's make the gestures a bit more noticeable to the user. Instead of just outputting to the console, let's provide something visible to the user:
> 
> * **DragGestureRecognizer** - Allow the user to drag AvatarView across the screen
> * **PinchGestureRecognizer** - Allow the user to resize the AvatarView height/width
> * **PanGestureRecognizer** - Allow the user to pan the AvatarView across the screen
> * **SwipeGestureRecognizer** - Allow the user to swipe the AvatarView across the screen
> * **TapGestureRecognizer** - Update the text in the AvatarLabel to the current tap count

@brminnick I didn't want to get dragged into this being an exercise in providing full sample code implementing absolutely everything.  I can/could spend more time on the gesture samples, but really feel that it is a much better use of our time and resources to invest in the actual 'proposal' control.  I personally think that the current 'gestures' sample gives the end consuming developer everything they need, and that we shouldn't be giving samples on how to implement what is "standard/common" MAUI functionality.
If you or any of the other CommunityToolKit gurus disagree, I will be more than happy to fully expand the gestures samples.

pictos commented on src/CommunityToolkit.Maui.Core/Primitives/Defaults/AvatarViewDefaults.shared.cs:
```diff
@@ -0,0 +1,36 @@
+using System.ComponentModel;
+
+/// <summary>Default Values for <see cref="IAvatarView"/></summary>
+[EditorBrowsable(EditorBrowsableState.Never)]
```
We don't like to have secrets on `Core`. So not sure if we want to hide this.

GeorgeLeithead replied:
> We don't like to have secrets on `Core`. So not sure if we want to hide this.

@pictos I was using `AlertDefaults` and `DrawingViewDefaults` as my guide.  These have the same _EditorBrowsable_ decorations.  I'll change to whatever is best?

brminnick replied:
Yup, George's implementation is totally valid and we should keep the `EditorBrowsable` attribute. 

Like `AlertDefaults`, these are internal defaults for our custom controls.

pictos commented on src/CommunityToolkit.Maui.Core/Interfaces/IAvatarView.shared.cs:
```diff
@@ -0,0 +1,73 @@
+﻿namespace CommunityToolkit.Maui.Core;
+
+/// <summary>Avatar view interface.</summary>
+public interface IAvatarView : IBorderView, ILabel, Microsoft.Maui.IImage, IImageSource
```
Interfaces are intended to be used for controls that will implement a `Handler`, which isn't the case of the `AvatarView` so we don't need the interface here, and since this control uses the `Maui.View` we don't need to have anything on `Core`, this can be all `Maui` specific.

Also, this interface has a lot of MVVM (the propertyChanged methods) smell and the `Core` project must not influence the implementor in any way. One more reason that we don't need it.

GeorgeLeithead replied:
> Interfaces are intended to be used for controls that will implement a `Handler`, which isn't the case of the `AvatarView` so we don't need the interface here, and since this control uses the `Maui.View` we don't need to have anything on `Core`, this can be all `Maui` specific.
> 
> Also, this interface has a lot of MVVM (the propertyChanged methods) smell and the `Core` project must not influence the implementor in any way. One more reason that we don't need it.

@pictos I was asked to move IAvatarView to `Core` in [This commen](https://github.com/CommunityToolkit/Maui/pull/481#pullrequestreview-1029242225).  Appreciate that the code has evolved, but I'm getting conflicting requests.  I'll change as to what is best, but not sure right now.

pictos replied:
We don't need to have an interface at all, as I said, interfaces are needed if the control will implement a `Handler`, the `AvatarView` is just a composition of Maui controls

brminnick replied:
The interface isn't necessarily a bad thing, because it'll help future toolkits like `CommunityToolkit.Comet` and `CommunityToolkit.BlazorWebView` implement their own AvatarView. 

But, I agree that it's a good idea to ensure it's not MVVM-specific. [Comet](https://github.com/dotnet/comet), for example, uses the MVU architecture. 

Does [Comet](https://github.com/dotnet/comet), for example, fire events like `OnCornerRadiusPropertyChanged()` and does it use methods like `IsTextSet()`? Or is that just an MVVM thing?

If other MAUI-based frameworks like Comet and [BlazorWebView](https://github.com/dotnet/maui/tree/main/src/BlazorWebView) don't leverage these method calls, then we should remove them from the interface.

pictos replied:
> The interface isn't necessarily a bad thing, because it'll help future toolkits like CommunityToolkit.Comet and CommunityToolkit.BlazorWebView implement their own AvatarView. (These future toolkits will also have a dependency on CommunityToolkit.Maui.Core).

I can see your point, and I'm on the fence. But if we keep the interface we also should avoid things like `Notes to implementer`, since we don't know where this will be implemented we shouldn't be restrictive. And this is why I don't like to use the interface in this case, we don't know which features other targets will provide, so this interface will be a good guide or a restrictive guide on how this `View` should be implemented?

> Does [Comet](https://github.com/dotnet/comet), for example, fire events like OnCornerRadiusPropertyChanged() and does it use methods like IsTextSet()? Or is that just an MVVM thing?

The `Changed` method is an MVVM  thing, about the `TextSet` I'm not sure but I would say it's an MVVM thing, MVU just `redraw` the UI. The interface doesn't need those `Changed` methods. 
Makes sense to have methods that describe an action (not the .NET one), for example, the [`IButton`](https://github.com/dotnet/maui/blob/main/src/Core/src/Core/IButton.cs) you have just methods that describe the `Pressed`, `Released`, and `Clicked` actions.

GeorgeLeithead replied:
Where I was going, for good or ill, I was trying to mirror as much from `Microsoft.Maui.Controls` as possible.  With a goal of making the control as easy to use for the end user as possible in it's simplest form, but still give the end user as much extensability as possible/feasible.

As such I took things like `Microsoft.Maui.Controls.ImageButton` as an example, and used this as my guide.  `ImageButton` inherits the interface `Microsoft.Maui.Controls.Internals.IButtonElement`, which has a seperation between properties and methods and includes **note to implementor**.  As such, I copied that, with the thought of "If it's included by Microsoft, then why shouldn't I?"

I do however totally get what @pictos is saying, in that including the note could be regards as restrictive, especially if this `view` is used as a guide going forward. Does the toolkit use `Microsoft.Maui` as a guide, or does it set it own?

VladislavAntonyuk replied:
I do not have objections to add interface. But in that case we also need t add it for other controls like UniformItemsLayout

pictos replied:
@GeorgeLeithead, I know that you followed the Maui structure, but there're some details (that we should probably document). Bear with me for a while.

If you see the [`IButton`](https://github.com/dotnet/maui/blob/main/src/Core/src/Core/IButton.cs) interface here, it's in the `Microsoft.Maui` namespace, and in the `Core` project, so it should be agnostic in terms of patterns. Now if you see the [`ITextAlignmentElement`](https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/ITextAlignmentElement.cs#L11) interface there's the `propertyChanged` method, but look! The namespace is `Microsoft.Maui.Controls` and it lives on the `Control` assembly, where the implementation isn't agnostic, in this case, is XAML and MVVM friendly.

On this project we followed this same idea, everything in the `Core` assembly should be agnostic in terms of implementation. And inside the `.Maui` assembly we are XAML and MVVM friendly.

I hope that makes things more clear to you, if don't feel free to asking questions, I'll be more than happy in provide all info that you need (:

GeorgeLeithead replied:
As everything I'm 'currently' doing IS XAML/MVVM friendly. Should I rename to `IAvatarViewElement` and move to `CommunityToolKit.Maui` namespace? (beside `AvatarView.shared.cs`). Then later if needed, Handler implementations could be added.

pictos replied:
There's no reason to have such an interface in the `.Maui` project. But I'll let @brminnick decide this, since he wants to have the `interface` in the shared layer to provide a guidance for other implementors.

pictos reviewed (dismissed):
Hey @GeorgeLeithead thanks for your contribution, I have some requests for you. Feel free to ping me if you need anything.

VladislavAntonyuk commented on samples/CommunityToolkit.Maui.Sample/Models/AvatarModel.cs:
```diff
@@ -0,0 +1,18 @@
+	public string Description { get; init; } = string.Empty;
+	public FontAttributes FontAttributes { get; init; } = FontAttributes.None;
+	public double FontSize { get; init; } = 18;
+	public double HeightRequest { get; init; } = 64;
```
Can we use SizeRequest so we ensure both Width and Height are set and have a valid value?

brminnick replied:
Let's keep `HeightRequest` and `WidthRequest` separate because it is easier to then use their values for XAML Bindings

VladislavAntonyuk commented on samples/CommunityToolkit.Maui.Sample/Models/AvatarModel.cs:
```diff
@@ -0,0 +1,18 @@
+
+public class AvatarModel
+{
+	public Color BackgroundColor { get; init; } = Colors.Black;
```
Can we make it dynamic depends on AppTheme?

brminnick replied:
Let's keep using `Colors.Black` because it works nicely with the default values for `BorderColor` (`Colors.White`) and `TextColor` (`Colors.White`).

VladislavAntonyuk commented on samples/CommunityToolkit.Maui.Sample/Pages/Views/AvatarView/AvatarViewBindablePropertiesPage.xaml:
```diff
@@ -0,0 +1,58 @@
+                    <Setter Property="LineBreakMode" Value="WordWrap" />
+                    <Setter Property="Margin" Value="4,0" />
+                </Style>
+                <Style x:Key="Divider" TargetType="BoxView">
```
From docs: Although BoxView can mimic simple graphics, a better alternative is to use .NET MAUI Shapes or .NET MAUI Graphics.

GeorgeLeithead replied:
This was simply copied directly from `ButtonPopup.xaml` and the other pages.  It's just used as a 'divider'.  Happy to change to something else you suggest.

brminnick replied:
If you can, let's replace it with a `Line`: https://docs.microsoft.com/dotnet/maui/user-interface/controls/shapes/line?WT.mc_id=dotnet-0000-bramin

GeorgeLeithead replied:
In `AvatarView*Page` all `BoxView` has been replaced with `Line`.

VladislavAntonyuk commented on samples/CommunityToolkit.Maui.Sample/Pages/Views/AvatarView/AvatarViewGesturesPage.xaml.cs:
```diff
@@ -0,0 +1,19 @@
+{
+	public AvatarViewGesturesPage(AvatarViewGesturesViewModel avatarViewGesturesViewModel) : base(avatarViewGesturesViewModel) => InitializeComponent();
+
+	void DragGestureRecognizer_DragStarting(object sender, DragStartingEventArgs e) => Debug.WriteLine("AvatarView drag gesture recognizer, drag starting.");
```
Let's use Toast to display the result.

Debug WriteLine is not visible on UI and user may assume the app doesn't work

GeorgeLeithead replied:
Great idea!
I have changed to use as short Toast instead of debug.

VladislavAntonyuk commented on samples/CommunityToolkit.Maui.Sample/ViewModels/Base/BaseGalleryViewModel.cs:
```diff
@@ -6,6 +6,14 @@ public abstract class BaseGalleryViewModel : BaseViewModel
 {
 	protected BaseGalleryViewModel(IEnumerable<SectionModel> items)
 	{
+		foreach (var item in items)
```
Iteration on IEnumerable leads to boxing and unnecessary memory allocation. Please use List or Array instead.

GeorgeLeithead replied:
My fork and branch are saying they are up to date, but this file (and others) isn't in-sync?  Do you have the GIT commands that I need to sync my fork and branch with the CommunityToolkit:main?

GeorgeLeithead replied:
I have converted to an Array, matching source, to prevent boxing and unnecessary memory allocation.

VladislavAntonyuk commented on samples/CommunityToolkit.Maui.Sample/ViewModels/Views/AvatarView/AvatarViewBindablePropertiesViewModel.cs:
```diff
@@ -0,0 +1,87 @@
+	{
+		new AvatarModel
+		{
+			Text = "SW", // Sam Worthington
```
Can the comment be in Description property?
Please add Description property in the sample as well

GeorgeLeithead replied:
The comment, and even the text used was just an amusement in that they are the actors' names from the 2009 movie "Avatar".  They don't add anything and could be any text at all (such as VA, BM, SL, PJ, GV), but I thought it might be regarded as an 'easter egg' for any user looking at the source code.

I have however added a description property to the model and bind them to the _SeamanticProperties_ in the sample view.

VladislavAntonyuk commented on src/CommunityToolkit.Maui.Core/Interfaces/IAvatarView.shared.cs:
```diff
@@ -0,0 +1,73 @@
+	/// <summary>Gets a value indicating the avatar border width.</summary>
+	double BorderWidth { get; }
+
+	/// <summary>Gets a value indicating the avatar corner radius (MAUI).</summary>
```
Do we need MAUI in summary?

GeorgeLeithead replied:
The .NET MAUI type `CornerRadius` is not standard across all controls that implement a corner radius (they instead currently use the type `Int`).  From what I can see it is **only** available on the BoxView control.  I have updated the comment to include a cref.

VladislavAntonyuk commented on src/CommunityToolkit.Maui.Core/Interfaces/IAvatarView.shared.cs:
```diff
@@ -0,0 +1,73 @@
+	/// <summary>Gets a value indicating the avatar corner radius (MAUI).</summary>
+	CornerRadius CornerRadius { get; }
+
+	//// NOTE to implementer: Implement these methods explicitly
```
Do we need it?
At least right now it is in not a correct place

GeorgeLeithead replied:
Been removed.

VladislavAntonyuk commented on src/CommunityToolkit.Maui.Core/Interfaces/IAvatarView.shared.cs:
```diff
@@ -0,0 +1,73 @@
+	//// NOTE to implementer: Implement these methods explicitly
+
+	/// <summary>Gets a value indicating the corner radius default.</summary>
+	CornerRadius CornerRadiusDefaultValue { get; }
```
Do we need it?
Can we set default value to CornerRadius property?

VladislavAntonyuk commented on src/CommunityToolkit.Maui.Core/Interfaces/IAvatarView.shared.cs:
```diff
@@ -0,0 +1,73 @@
+	CornerRadius CornerRadiusDefaultValue { get; }
+
+	/// <summary>Gets a value indicating the text default.</summary>
+	string TextDefaultValue { get; }
```
Should it be nullable?

GeorgeLeithead replied:
Made nullable.

VladislavAntonyuk commented on src/CommunityToolkit.Maui.Core/Interfaces/IAvatarView.shared.cs:
```diff
@@ -0,0 +1,73 @@
+
+	/// <summary>Padding default value creator method.</summary>
+	/// <returns>Padding thickness.</returns>
+	Thickness PaddingDefaultValueCreator();
```
Should it be in interface?
I would extract it to another interface or make a private method in AvatarView class

GeorgeLeithead replied:
I have added `IPaddingElement`, added a default to `AtarViewDefaults` and updated the method to use these both. In `Microsoft.Maui.Controls.Border`, to which this control inherits, this method is _public_.

GeorgeLeithead replied:
Removed from interface as now unnecessary.

VladislavAntonyuk commented on src/CommunityToolkit.Maui.Core/Primitives/Defaults/AvatarViewDefaults.shared.cs:
```diff
@@ -0,0 +1,36 @@
+public static class AvatarViewDefaults
+{
+	/// <summary>Default avatar border width.</summary>
+	[EditorBrowsable(EditorBrowsableState.Never)]
```
Do we need it on each property if we applied it to class?

GeorgeLeithead replied:
I mirrored what was done for `DrawingViewDefaults.shared.cs`.  More than happy to change if it's the right thing to do?

VladislavAntonyuk commented on src/CommunityToolkit.Maui/Views/AvatarView.shared.cs:
```diff
@@ -0,0 +1,446 @@
+namespace CommunityToolkit.Maui.Views;
+
+/// <summary>Avatar content view.</summary>
+public class AvatarView : Border, IAvatarView, IFontElement, ITextElement, IImageElement, ITextAlignmentElement, ILineHeightElement, ICornerElement
```
I think all .NET MAUI interfaces should be parents for IAvatarView

pictos replied:
Some of them can't be, because they live in the `Maui.Controls` namespace, e.g. the [`IImageElement`](https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/IImageElement.cs)

GeorgeLeithead replied:
All the `Microsoft.Maui` interfaces were moved to the `IAvatarView`.  All the `Microsoft.Maui.Controls` interfaces can't be because of their namespace

GeorgeLeithead commented:
If not already discussed, could I suggest that we add `StyleCop.Analyzer` to the various projects to encourage rule compliance, like with `.editorconfig`?  Some rules are always going to be contentious (such as "SA1101: Prefix local calls with this", which can be configured to alert 'none').

GeorgeLeithead commented:
> ### AvatarImage Doesn't Clip to AvatarView Bounds
> I noticed that the AvatarImage sticks out past the border of the AvatarView.
> 
> Let's make sure the image stays inside the bounds of the border.
> 
> <img alt="Screen Shot 2022-07-15 at 7 14 36 PM" width="1360" src="https://user-images.githubusercontent.com/13558917/179336447-c0ed0b80-0150-4df1-a474-cf349cd33e44.png">

@brminnick Image clipping for a `Border` control works OTB on `Android`.  I have a working resolution for `Windows` (I'm not 100% happy with it, so please have a look).  I am not able to verify on iOS or MacCatalyst as I don't have any of these platforms.

<image src="https://user-images.githubusercontent.com/43666574/180270833-3f21ca16-0c50-4611-b46a-f038d17a5630.png" width=250 alt="Android screenshot" />
<image src="https://user-images.githubusercontent.com/43666574/180270886-c0be29b2-8ce1-4b56-a99b-f760ff2fb880.png" width=250 alt="Windows screenshot" />

brminnick reviewed (approved):
Thanks @GeorgeLeithead!!! Approved!!! 🚀 🙌 

FYI - I tweaked the Image Clipping logic a bit to ensure it works on all platforms and takes the `BorderWidth` into account

brminnick commented:
FYI to all - there is a bug on iOS for `Microsoft.Maui.Controls.Border`: when setting the `BackgroundColor` property, iOS paints the color on top of the `Border.Content` (i.e. the color is painted on top of the avatarImage on iOS).

I've implemented a workaround and noted in the code here:
https://github.com/GeorgeLeithead/CommunityToolKitMaui/blob/29cde2dd3a2c625007774d7413ca32fb39d9ae32/src/CommunityToolkit.Maui/Views/AvatarView.shared.cs#L324-L341

Here's an example of what it looks like (`BackgroundColor` is set to `Colors.Black`): 
![image](https://user-images.githubusercontent.com/13558917/180352566-088180f4-e99c-4ff2-9d00-c00a1010dedb.png)
